src/main/java/org/openstreetmap/atlas/checks/validation/tag/RoadNameSpellingConsistencyCheck.java

```diff
@@ -0,0 +1,101 @@
+package org.openstreetmap.atlas.checks.validation.tag;
+
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.openstreetmap.atlas.checks.base.BaseCheck;
+import org.openstreetmap.atlas.checks.flag.CheckFlag;
+import org.openstreetmap.atlas.geography.atlas.items.AtlasObject;
+import org.openstreetmap.atlas.geography.atlas.items.Edge;
+import org.openstreetmap.atlas.tags.HighwayTag;
+import org.openstreetmap.atlas.utilities.configuration.Configuration;
+import org.openstreetmap.atlas.utilities.scalars.Distance;
+
+/**
+ * Flags road segments that have no {@link org.openstreetmap.atlas.tags.names.NameTag} or a NameTag
+ * with a different spelling from other segments of the same road. This check is primarily meant to
+ * catch small errors in spelling, such as a missing letter or letter accent mixups.
+ *
+ * @author seancoulter
+ */
+public class RoadNameSpellingConsistencyCheck extends BaseCheck<Long>
+{
+
+    private static final long serialVersionUID = 1L;
+
+    private static final HighwayTag MINIMUM_NAME_PRIORITY_DEFAULT = HighwayTag.SERVICE;
+    private static final double MAXIMUM_SEARCH_DISTANCE_DEFAULT = 300.0;
+    private static final double MAXIMUM_CHARACTER_DIFFERENCES_DEFAULT = 1;
+    private static final String FALLBACK_INSTRUCTIONS = "These road segments have spelling inconsistencies. One spelling is: %s. Examine all flagged road segments to determine the best spelling, and apply this spelling to all of those segments.";
+
+    private final Distance maximumSearchDistance;
+    private final double inconsistentCharacterCountThreshold;
+
+    public RoadNameSpellingConsistencyCheck(final Configuration configuration)
+    {
+        super(configuration);
+        this.maximumSearchDistance = this.configurationValue(configuration,
+                "distance.search.maximum", MAXIMUM_SEARCH_DISTANCE_DEFAULT, Distance::meters);
+        this.inconsistentCharacterCountThreshold = this.configurationValue(configuration,
+                "characterCount.inconsistencies", MAXIMUM_CHARACTER_DIFFERENCES_DEFAULT);
+    }
+
+    /**
+     * This function will validate if the supplied atlas object is valid for the check.
+     *
+     * @param object
+     *            the atlas object supplied by the Atlas-Checks framework for evaluation
+     * @return {@code true} if this object should be checked
+     */
+    @Override
+    public boolean validCheckForObject(final AtlasObject object)
+    {
+        return object instanceof Edge
+                && ((Edge) object).highwayTag().isMoreImportantThanOrEqualTo(
+                        MINIMUM_NAME_PRIORITY_DEFAULT)
+                && ((Edge) object).isMasterEdge() && !this.isFlagged(object.getIdentifier())
+                && ((Edge) object).getName().isPresent();
```

Comment:
I commented this on @Bentleysb's comment, but it's already been resolved so it's hidden. Take two:
First of all, respect for this! Love to see dynamic programming in the wild 🙌

I would love to see this pulled out into a utility somewhere. In addition to making it public for others to use, that allows us to write unit tests for it which will increase our confidence in this covering all of the edge cases.

As another thought, do you see the threshold for this ever increasing above 1? Or does it not make sense to ever check for a minimum edit distance of 2, for example?

Reply:
I like the idea of this being a utility. I can see it being useful for at least a few cases when we're dealing with tags. I'd have to do some research and then I think we can get the discussion going on how exactly we want to implement it.

Regarding your question, I don't see the threshold increasing above 1. The task for this problem is to catch "smaller" issues in spelling (accent mixup, one letter missing/added, etc) which I believe are all covered by an edit distance of 1. However I do think we should allow the edit distance to be bumped up if we pull out a utility. If we decide to go that route, I suggest we get deeper test coverage, since an increased edit distance brings substantially more corner cases/things to watch out for.